### ISSUE TITLE
Better automatic player color assignment

### ISSUE THREAD

**Issue by pi99y** [OpenRA/OpenRA#11248]
The automated color assigned to players (mostly a problem with AI players / skirmish) usually needs colors to be adjusted manually as the automatic picks can be pretty indistinguishable:

![automatic_color_asigned](https://cloud.githubusercontent.com/assets/10639257/15106933/d555f86c-15cc-11e6-9add-11f02ed72ad6.jpg)

My suggestion when picking an automated color is to:
- befiore repeating a hue pick colors from different hues - don't add another green when other hues have not been used yet
- first suggest colors that are brighter and more saturated - they are easier to distinguis from one another (colors located roughly in the upper right side of the color picker)

![automatic_color_asigned_tones](https://cloud.githubusercontent.com/assets/10639257/15107028/88f3b8a0-15cd-11e6-9ee5-86e7ffea96f5.jpg)

**Comment by pizzaoverhead** [comment]
Rather than randomise and police what colours are chosen, we could specify a selection of mod-specific colours which are known to look good. For example, the original games each had a preset number of colours that could be used. See the top right here:

![RA95 skirmish lobby](http://www.mobygames.com/images/shots/l/64900-command-conquer-red-alert-the-aftermath-windows-screenshot.gif)

**Comment by pi99y** [comment]
I'm all for the pre-set colors @pizzaoverhead but I was under the impression that many players wanted the color picker and not the pre-set colors...

**Comment by pizzaoverhead** [comment]
This would be for automatic colours only. You could still customise them however you wanted.

**Comment by pchote** [comment]
@pizzaoverhead do you want to generate a list of ~20 sensible colors for each mod?  I could then include them as part of the fix for #10906.

**Comment by pi99y** [comment]
Also please consider choosing colors that contrast the ore, gems and terrain (grass, snow, sand, water) as some smaller units (rifle infantry etc...) can be hard to spot if they are in a color that's too similar to its surrounding. Thanks!

**Comment by pizzaoverhead** [comment]
@pchote Sure, I'll put something together.

**Comment by pizzaoverhead** [comment]
So far I've grabbed the original multiplayer colours from the three main mods in RGB format:
```
D2K
107, 130, 165   Grey teal
214, 81, 74     Light red
107, 162, 132   Dull green
123, 16, 140    Purple
132, 56, 16     Brown
173, 136, 8     Dark gold

RA
196, 176, 96    Copper
68, 148, 228    Blue
236, 0, 0       Red
0, 196, 0       Green
252, 136, 0     Orange
112, 112, 112   Grey
68, 144, 124    Teal
112, 44, 36     Burgundy

TD
252, 252, 84    Yellow
244, 0, 0       Red
0, 112, 112     Teal
212, 120, 16    Orange
84, 252, 84     Green
164, 164, 188   Grey
```

**Comment by pi99y** [comment]
Thank you @pizzaoverhead for doing this.

Here's an image of how those colors look in sRGB.
(BTW: Do the ingame colors change maybe? I know I get a bit different colors in my screen grabs of the game compared to how they look in game ... but nothing drastic)

![openra_colors](https://cloud.githubusercontent.com/assets/10639257/15427140/408b3ee8-1e93-11e6-8172-ad5e7a1e34b0.png)

By looking just at RA colors - are the "copper" and "teal" mybe a bit too close to the terrain (ore and grass)? Perhaps a bright yellow and pink as replacements for the two?

**Comment by StexoO** [comment]
If you use predefined colors, what do you do if a map has more spawns than the number of predefined colors ?
For example if you play a ten players map in RA, the last two players still have a completely random color instead of a predefined one when they connect to the server ?

**Comment by GraionDilach** [comment]
There was an idea of having X amount of predefined colors and Y amount of customizable slots per plpayer. I do hope this latter is still planned otherwise the revamp is a clear step backwards.

**Comment by pchote** [comment]
This ticket is about the initial automatic colours that are assigned, and does not impact the UI in any way.  If there are more players than colors then it would fall back to its current method, but like I mentioned above I would like to have at least 20 decent colors to justify the effort of implementing this (its a waste of time if it can't even cope with our biggest shipped maps).

**Comment by pi99y** [comment]
I feel 20 distinguishable colors is really hard to achieve.

Here's my try of 20 colors. I tried to simulate the worst - small unit (rifle infantry) on different terrain. 

![ra_colors](https://cloud.githubusercontent.com/assets/10639257/15496357/00a9314a-2195-11e6-9712-7adf78654a3e.png)

When automated color selection takes place it should still have an order in which it picks these colors, so that it will only take a second red when other hues are exhausted.

**Comment by pi99y** [comment]
Maybe in an order *like* this (left to right): 
![ra_colors_order](https://cloud.githubusercontent.com/assets/10639257/15497157/87495cda-2199-11e6-9c73-5af2ad99b264.png)

**Comment by pi99y** [comment]
How can I add **fixed player color** to map.yaml **with skirmish maps?**
I'd like to further test (and adjust if needed) these colors in-game.
Thanks.

**Comment by GraionDilach** [comment]
https://github.com/OpenRA/OpenRA/blob/bleed/mods/cnc/maps/cnc64gdi01/map.yaml#L35

**Comment by pi99y** [comment]
Thanks @GraionDilach I was able to achieve what I wanted with this. I'll post screens/colors as soon as I come up with it.

**Comment by pi99y** [comment]
So here is my new ordered list with in-game screenshots:
![openra_colors_select_pi99y](https://cloud.githubusercontent.com/assets/10639257/15504850/42630544-21c1-11e6-9b66-2969ac72ff5a.png)

Automated colors would be picked in order 1 - 20 from the list and skipped if too similar to hand-picked/default color of a human player.
What do you think?

**Comment by GraionDilach** [comment]
10's tone should also be either darkened either 6 should be more desaturated, because they do look similar to me. Or shift 19 into 6 and use a bit darker tone of the Munsell red - https://en.wikipedia.org/wiki/Shades_of_red#Red_.28Munsell.29 - (something along HSL 345-100%-35% or RGB 179-0-45) for the vacant slot. The distance between 8 and 18 should also be increased (by darkening/desaturating 18 or lightening 8 up).

It looks fine for the most part though.

**Comment by GraionDilach** [comment]
Taking another look, 4-20 and 7-12 are also a pair which I'd say need to be differed. 12 prolly can be hueshifted into reddish with 5-10° without an issue (it's already the darkest color you picked), while for 20, maybe replace it with a color like RGB 170-170-85.

**Comment by pi99y** [comment]
Hi @GraionDilach 
Sorry, I was just about to post this new one and then I saw you comment, so this does not yet reflect your comments...

But since it's quite time consuming for me to make these. Can you please take another look at the new one and tell me which of your suggestions still apply.
Thanks.

![openra_colors_select_pi99y2](https://cloud.githubusercontent.com/assets/10639257/15509046/516df574-21d2-11e6-9441-6761a24a49a0.png)

**Comment by GraionDilach** [comment]
1-17 still needs tweak. Everything excluding 7-12 from my previous comments still apply as well.

**Comment by pi99y** [comment]
OK thanks, I'll make a new version - but probably not before tommorow morning (in ~15 hours).

**Comment by pi99y** [comment]
OK, so I tried to address the things you pointed out @GraionDilach. I eventually ended up with so many corrections and swaps, that I just corrected it again based on my observation ... IDK, I really hope it's better. I've pushed the colors I feel are less unique to the end of the list so they will get used less frequently.

at 100%
![ra_colors_pi99y_v3](https://cloud.githubusercontent.com/assets/10639257/15537332/0c64c3d2-2276-11e6-93e6-fd9bf82a4570.png)

at 200%
![ra_colors_pi99y_v3double](https://cloud.githubusercontent.com/assets/10639257/15537347/1cee6c26-2276-11e6-9132-8a5eeb2e6bc6.png)

```
color codes:
01 #91d2ff
02 #ff261f
03 #75f107
04 #ffffff
05 #fdff0e
06 #ce00ff
07 #1a20ff
08 #ff7e00
09 #90ffad
10 #b90258
11 #000000
12 #ff91f6
13 #00a2ff
14 #fff7c3
15 #02b3a5
16 #098e00
17 #a9a9a9
18 #ccacff
19 #5a5cab
20 #c79679


```

**Comment by pi99y** [comment]
As a reference for further color validator improvements:
* darker colors (on samall units) are harder to distinguish, especially on snow (light background)
* some light colors (like white) work good even on snow

Current validator bans white but allows many dark colors.

**Comment by GraionDilach** [comment]
Isn't white used by the tech buildings?

4-12 is a problem, 9-13-15 as well. Why did you rejected my #AAAA55 suggestion?

**Comment by pi99y** [comment]
I'll look in to it some more. I didn't reject anything but one change led to another and so on and I just wasn't able to apply your comments to the constantly changing version... Now that this is hopefuilly a better starting point it will maybe be easier. I'll report back with a new version.

BTW what kind od monitor are you viewing this on? Combinations you mention are quite clear here on two different screens. One callibrated higher-end one and one a cheap smart phone. I'm not saying they don't need change, I just don't want to go in the wrong direction. I know differences should be visible on standard monitor and as far as I can tell I did everything (soft proofing and trying a differnet, cheaper screen) to combat that. Does the difference between those colors becomes any clearer if you fiddle with your setting, like brightness?

Anyway I'll do some more color fiddling and post updates :)

**Comment by pi99y** [comment]
Just to be more sure: Can you please confirm what you're seeing on another screen? Thanks.

**Comment by GraionDilach** [comment]
ASUS VW221D - http://www.cnet.com/products/asus-vw221d-lcd-monitor-21-6-series/specs/  - preset Standard mode. Although I went through the presets now and Standard seems to desaturate everything a bit, although the others seemingly have biased-toward-a-particular-color setups. Infact, the only time I can't see 4-14 a problem, I have to switch to Theater Mode (where I have controls over saturation) and have to increase saturation so that Github comment button becomes glowingly bright green, etc.

**Comment by GraionDilach** [comment]
Okay, rechecked it on my sister's ASUS VW195 (Noidea how that monitor is set up though). Only 4-14 is a problem, the 9-13-15 apparently blends due to ym preset.

**Comment by pi99y** [comment]
Thanks for testing it. I will continue with this tomorrow + check two other displays to have a better understanding of what is acceptable for most setups.

**Comment by pi99y** [comment]
@GraionDilach This (B) is the same image and color combination however I have not color managed it to sRGB. Do you see it any differently than the previous version (A) and if so - in which version (A or B) you think that the background (desert, snow, grass, ore) looks as it looks to you in game? Thanks.

A: (previous)
![ra_colors_pi99y_v3_srgb](https://cloud.githubusercontent.com/assets/10639257/15553370/f50769c6-22be-11e6-95b2-2caab5271282.jpg)

B: (new)
![ra_colors_pi99y_v3](https://cloud.githubusercontent.com/assets/10639257/15553264/8868f456-22be-11e6-96c0-836346c29eb5.jpg)

**Comment by pi99y** [comment]
OK, so here's the new (let's call it version 4) selection. I've pushed the more problematic colors to 16-20 where they should see minimal use.

Is it any better @GraionDilach?
![ra_colors_pi99y_v4](https://cloud.githubusercontent.com/assets/10639257/15555264/41d7db74-22c8-11e6-8eb2-261e98e98b61.jpg)

Also, @pizzaoverhead @pchote  @abcdefg30  @StexoO (and others) can I get your opinion too? It's possible that me and @GraionDilach are getting blind from all this pixel spotting...
Thanks.

**Comment by pchote** [comment]
The initial color selection would be randomized, so order doesn't matter.

**Comment by pi99y** [comment]
@pchote But that's a shame since obviously some colors work better (in all conditions) than others ... Why the need for randomization? User will still be able to manually adjust colors if he really wants something specific...

**Comment by pchote** [comment]
Selecting the same player colors over and over again is repetitive and lame. If a color is a bad choice then it should not be included in the initial set.

**Comment by pi99y** [comment]
@pchote OK, but I think that this "exercise" shows that we can get combinations of maybe 8-10 colors that are at the same time distingushable and work well in all condistions. I think 20 is a bit unrealistic. Apparently on some monitors/settings this is even more so (see @GraionDilach's findings about his monitors above).

How about a compromise, first 10 colors get randomized then others follow in order?

**Comment by pchote** [comment]
I'm not prepared to compromise – either this feature works well and is worth adding, or it doesn't and is not worth the time and complexity to implement.

**Comment by pi99y** [comment]
This feature can work well (and as you intended) if we agree how many colors (min) is acceptable.

For now I'm just asking you all if you think that my selection of colors is any good and how many of them do you personally think are destinguishable enough. Maybe most will be happy with all 20 colors. 

I think that the first 15 could be acceptable and not because I didn't try hard enough for 20, it's just a bit too much.

So let's hear what you all think of the selection and how/if it can made better.

**Comment by pchote** [comment]
20 is just an arbitrary number.  If there are only 10 good distinct colors, then use 10 and let the rest be randomly picked.

**Comment by StexoO** [comment]
All the colors in https://github.com/OpenRA/OpenRA/issues/11248#issuecomment-221698423 are pretty different and I see them pretty well on my screen. Even colors like 3-9 or 4-9 are sufficiently different and colors like 5-14 can be seen on ore (I think the rifle helps too).
:+1:
Maybe 90 % of the maps available in the resource center don't have more than 10 (or 20) players. It could be an interesting statistic.

**Comment by GraionDilach** [comment]
I still feel 4-14 blending together and easy to be mixed when both are present. The rest are fine.

**Comment by abcdefg30** [comment]
I think @GraionDilach is right, at least for snow maps. Otherwise I'm with @StexoO, looks good to me. ✅

**Comment by pi99y** [comment]
@GraionDilach @abcdefg30 Since I can see the 4-14 difference quite clearly on all of my screens (4 different), can you please guide me into which direction to change the no. 14 (currently: #fff5c5)?

Thanks.

I feel like only keeping the first 15 if 20 is not necessary. You OK with that?

**Comment by GraionDilach** [comment]
Could you upload the maps? I guess I shouldn't shoot in the dark.

**Comment by pi99y** [comment]
sure: http://filebin.net/mn90czhb2o/ra_player_color_maps_test.zip

**Comment by StexoO** [comment]
@pi99y I got this

Players count <= | Maps count | %
---|---|---
0 | 28 | 0.58
1 | 158 | 3.25
2 | 818 | 16.83
3 | 965 | 19.86
4 | 2338 | 48.11
5 | 2422 | 49.84
6 | 3239 | 66.65
7 | 3299 | 67.88
8 | 4233 | 87.10
9 | 4293 | 88.33
10 | 4502 | 92.63
11 | 4534 | 93.29
12 | 4644 | 95.56
13 | 4648 | 95.64
14 | 4750 | 97.74
15 | 4754 | 97.82
16 | 4790 | 98.56
17 | 4792 | 98.60
18 | 4805 | 98.87
20 | 4836 | 99.51
22 | 4844 | 99.67
25 | 4846 | 99.71
28 | 4848 | 99.75
30 | 4852 | 99.84
36 | 4856 | 99.92
50 | 4858 | 99.96
100 | 4860 | 100.00

**Comment by pi99y** [comment]
Great, so 15 colors covers ~98% of maps. That's pretty good. Thanks.

**Comment by pi99y** [comment]
So, dropped to 15 colors and adjusted #14.
![ra_colors_pi99y_v4 1](https://cloud.githubusercontent.com/assets/10639257/15626735/19c41920-24cd-11e6-8f15-fe2d94961cd9.jpg)

If everything is OK now, I'll move now to selecting the TD mod pallete, OK?

**Comment by GraionDilach** [comment]
Looks good. IMO these should work well as-is for the rest too.

**Comment by pchote** [comment]
I just noticed that you have pure white and pure black as two of your colors.  Those are outside the allowed selectable color ranges, and will be rejected by the validator.

**Comment by pi99y** [comment]
@pchote  Yeah I commented on that in a comment above. These two colors work well, so maybe the validator needs to pass them trought?

Otherwise I think I'll just rather drop to 12-13 colors instead.

BTW: here's the same palette for TD. I think this might need finetuning per mod. What do you say?
![td_colors_pi99y_v1](https://cloud.githubusercontent.com/assets/10639257/15626940/f7cff580-24d3-11e6-8d69-64c1bb973433.jpg)

**Comment by GraionDilach** [comment]
That's stupid. You shouldn't go down below 15. Drop white and black instead. Color 16 and 20 from your last 20slot would do in place of them.

**Comment by pi99y** [comment]
OK, I'll come up with something. In the meantime... what do you think about this palette in TD?
Smaller infantry = bigger problems :) Also The more colorful tiberium...
But it is true that putting infantry in tib is less common since it burns...
idk..

**Comment by pi99y** [comment]
Also @pchote  are white and black def. being banned or is there chance for adjusting the validator. I mean, the test thows they are not problematic colors... Thanks

**Comment by pchote** [comment]
Colors defined directly on the player references are not validated at all.  White in particular is problematic because that is the color of the neutral actor.  Some of the other colors (like that blue) looks like they may be too close to other terrain colors to be accepted.

The color validator works by calculating distances between colors using a pseudo-human-perception-like distance scale, it's not possible to simply white-list specific colors.

**Comment by pi99y** [comment]
How do you suggest I test if they are accepted by the validator - I have color codes to enter and that's not supported by the UI?

**Comment by pi99y** [comment]
So here's a palette that actually **validates**:
![ra_colors_pi99y_v4 3](https://cloud.githubusercontent.com/assets/10639257/15634338/946166b6-25c1-11e6-94b5-fba25e7ada2b.jpg)
Passable?

**Comment by pi99y** [comment]
This is  the SAME PALETTE **but reordered** to group similar colors to see if they are distinguishable. Fingers crossed :)

![ra_colors_pi99y_v4 3_reordered](https://cloud.githubusercontent.com/assets/10639257/15634426/8d011162-25c3-11e6-9f1a-37e9a446d873.jpg)

**Comment by pi99y** [comment]
Are the validator restrictions the same for ra, td, d2k or do I have to test it for each mod separately?

**Comment by pchote** [comment]
The validator restrictions are based on the tileset colors.

**Comment by pi99y** [comment]
Hopefully the last RA colors version. :)
Please, if you have any issues with it comment now, as I plan to move on to other two mods tommorow.
![ra_colors_pi99y_v4 5](https://cloud.githubusercontent.com/assets/10639257/15635442/b4900d1c-25de-11e6-9c55-976097c88351.jpg)
Color codes for current selection: 1 #ffc9ca, 2 #f50606, 3 #98331f, 4 #f57606, 5 #f7bb06, 6 #f861a4, 7 #da06f3, 8 #ddb8ff, 9 #06f739, 10 #cef7b2, 11 #200738, 12 #280df6, 13 #2f86f2, 14 #76d2f8, 15 #34ba93

**Comment by pi99y** [comment]
OK, now the D2k colors: 
* based on RA colors but with more green colors (since it's absent in the tileset)

![d2k_colors_pi99y_v1](https://cloud.githubusercontent.com/assets/10639257/15642111/81d137fe-2645-11e6-9150-4128a4181f40.jpg)

Color codes:
1 #ffc9ca, 2 #f53333, 3 #ffae00, 4 #fff830, 5 #87f506, 6 #f872ad, 7 #da06f3, 8 #ddb8ff, 9 #def7b2, 10 #39c46f, 11 #200738, 12 #280df6, 13 #2f86f2, 14 #76d2f8, 15 #498221

Any comments? Please look at the color/s in question on all backgrounds since it usually looks different on a different tileset and there has to be a compromise between all of them.

@pchote While I managed to check that single colors pass the color validator I cannot check the "between colors difference" validation since I can enter the color code just for one player. Colors defined in map.yaml don't seem to get validated. Can you please help me verify that the posted RA and D2K colors meet that validation? Thanks.

**Comment by pi99y** [comment]
TD colors:
* these are the hardest to pick since infantry is smaller, tiberium is more colorful 
* I'd argue that while some colors are similar to the tiberium (problematic for infantry), infantry dies in tiberium and those colors are not problematic on other terrain or bigger units

(x2 and x1 zoom, (green bg tiles) jungle and winter were so similar that i included just one in the preview)
![td_colors_pi99y_v1](https://cloud.githubusercontent.com/assets/10639257/15667916/015371bc-2719-11e6-9d6d-711c68ece36d.jpg)

Color codes:
1 #f70606, 2 #ff7a22, 3  #f8d3b3, 4 #f8e947, 5 #b6f706, 6 #f335a0, 7 #a64d6c, 8 #ce08f9, 9 #f5b2db, 10 #12b572, 11 #380135, 12 #1d06f7, 13 #328dff, 14 #78dbf8, 15 #cef6b1

**Comment by pi99y** [comment]
If noone comments on these picks I consider the 3 last palettes (RA, D2k, TD) done for now.
It would be great if someone could implement it now (@pchote?). Thanks.

**Comment by abcdefg30** [comment]
Sorry for the late response.

I think the colors are not quite perfect yet. Although I can see the differences of all colors if they're plain next to each other, following are not clearly distinguishable ingame (on terrain):
- In RA and D2k: 7 - 13
- In TD: 8 - 13 (not on temperat though)

Anyway, two things:
1) I might not be the best person to ask, as I'm a bit colorblind, so above should be confirmed.
2) Great work! I really like the effort you're putting into this issue, especially with screenshots on different terrains.

**Comment by pi99y** [comment]
Hi @abcdefg30 thanks for your response and suggestion. Can you please refer some more people to give their opinion here since the 7-13 and 8-13 are very clear (strong/~toxic purple vs sky blue) to me here on different screens even, so it's hard for me to respond with a right change. I even tried simulating a restricted srgb monitor when I was picking the colors but if anything the red-orange colors were closer together than these two. I realize colors can be quite subjective and that's why we really need more people to look into this. 

I'll do my best to try new combos but I would like to see more opinions to know what is an issue for most people.

**Comment by obrakmann** [comment]
Those particular colors look fine to me. Just chalk it up to @abcdefg30's disability :p

(@pi99y, you might have made this harder for yourself by testing this on infantry instead of, say, a 1tnk.)

**Comment by pi99y** [comment]
Yeah I wanted a ~worst case scenario. So if these pass I'm pretty confident it will work well in practice, at least most of the time :)

**Comment by swinbank** [comment]
There are some algorithmic approaches to this which could work with all mods and tile sets.

Some solutions are more robust than others, but I'll detail one which involves changing the color picker UI, and integrating the validator's responsibility into it. This would involve some work, which I don't personally have time to do, unfortunately. It was a fun problem to think about nonetheless. Hopefully some of these ideas serve useful.

The general strategy is to turn this into a geometry problem.

**Step 1**
Unify the color picker's 1D (hue) + 2D (saturation, brightness) color space into a single plane. There's no perfect way to do this since color spaces naturally take various 3D forms. For the purpose of the OpenRA color picker, I think a good-enough 2D projection would do. The examples below are actually terrible in that they are perceptually non-uniform, but you get the idea. It could be a rectangle, a circle, a diamond, whatever works. The picker's UI would benefit from being larger, to accommodate the increased information density.

![image](https://cloud.githubusercontent.com/assets/3004626/17650992/eee40962-6221-11e6-807f-292be1667607.png) 
![image](https://cloud.githubusercontent.com/assets/3004626/17651226/e50261ae-6227-11e6-82e1-993b0000b580.png)


**Step 2**
Mark the coordinates of each reserved color, like those computed from the tile set and ones that have been manually chosen by the players.

**Step 3**
Compute a voronoi of the coordinates, the vertices of which then represent the coordinates of colors that are the farthest away from the reserved colors, and thus are the most perceptually different*.

![image](https://cloud.githubusercontent.com/assets/3004626/17651012/81e4c09e-6222-11e6-90a6-52303fbd7f4f.png)

**Step 4**
Among the set of colors (vertices) computed from the voronoi, assign them randomly to the players. It would probably be good to weight the selection using some heuristics. Also, you'd probably need to recompute the voronoi as the number of available vertices drops below some threshold. Actually, I suppose you'd want to recompute it after every new color pick.

**Step 5 (bonus)**
For each reserved color, like the ones computed from the tile set, and for each player color, draw an ellipse around it representing the exclusion zone where other colors cannot be picked. The ellipses can overlap, but the centroid of each cannot intersect another ellipse. Depending on how the color space was projected in step 1, the ellipses would need to be scaled corresponding to the perceived similarity of nearby colors. In the example below, the ellipses would be enlarged as they approach the "poles" of pure black or white, since the colors in those regions start to look similar. Around the equator, they'd look closer to circles. Any way, when allowing a player to pick a color by hand, simply reject clicks that intersect any of the ellipses. This would obviate the trial-and-error tango with the color validator.

![swiss_cheese](https://cloud.githubusercontent.com/assets/3004626/17650787/45889c3e-621c-11e6-98c0-5691fcbee85a.jpg)

----

**Footnote:* There are some color spaces that are more ideal than others for this type of thing. The [HCL color space](http://vis4.net/blog/posts/avoid-equidistant-hsv-colors/), for example, is a "perceptually uniform" color space, where the distance between two colors is roughly equal to their perceptual difference. This is unlike the HSB/HSV/HSL spaces that are ingrained in us. Even if you were to hardcode the first N "random" colors for players, it might be a good idea to use an HCL color picker (you can find them on the web) and try to adhere to some average distance between the points.

**Comment by pi99y** [comment]
Sorry for my inactivity. 
@swinbank I really like your idea with ellipses to show the colors that are too similar that should be a part of GUI!
 
While a better algorithm for picking colors is surely the way to go long term I think my "list of good starter colors" is still valid for automated picks and could be a good quick fix until a more complex solution is ready. 
 
If anyone wants to implement that - the color picks (above posts) are finalized as far as I'm concerned. Thanks.

**Comment by soft-war** [comment]
Is it bad to necro a thread like this in Github? This seems like a case where pi99y and others have gotten this 95% figured out. Would it be bad to implement this now?

**Comment by pchote** [comment]
Fixed by #14853!
